packages/reaction-core/client/subscriptions.js

```diff
@@ -3,36 +3,44 @@
  * Create persistent sessions for users
  * The server returns only one record, so findOne will return that record
  * Stores into client session all data contained in server session
- * supports reactivity when server changes the serverSession
+ * supports reactivity when server changes `newSession`
  * Stores the server session id into local storage / cookies
+ *
+ * Also localStorage session could be set from the client-side. This could
+ * happen when user flush browser's cache, for example.
+ * @see https://github.com/reactioncommerce/reaction/issues/609#issuecomment-166389252
  */
-let currentSession;
-let serverSession = Random.id();
-
 Tracker.autorun(function () {
-  currentSession = Session.get("sessionId") || amplify.store(
-    "ReactionCore.session");
-  if (!currentSession) {
-    amplify.store("ReactionCore.session", serverSession);
-    Session.set("sessionId", serverSession);
-    currentSession = serverSession;
+  // we are trying to track both amplify and Session.get here, but the problem
+  // is - we can't track amplify. It just not tracked. So, to track amplify we
+  // are using dirty hack inside Accounts.loginWithAnonymous method.
+  const sessionId = amplify.store("ReactionCore.session");
+  let newSession;
+  Tracker.nonreactive(() => {
+    newSession = Random.id();
+  });
+  if (typeof sessionId !== "string") {
+    amplify.store("ReactionCore.session", newSession);
+    Session.set("sessionId", newSession);
+  }
+  if (typeof Session.get("sessionId") !== "string") {
+    Session.set("sessionId", amplify.store("ReactionCore.session"));
   }
+  ReactionCore.Subscriptions.Sessions = Meteor.subscribe("Sessions",
+    Session.get("sessionId"));
 });
 
-ReactionCore.Subscriptions.Sessions = Meteor.subscribe("Sessions",
-  currentSession);
-// Load order is important here, sessions come before cart.
-ReactionCore.Subscriptions.Cart = Meteor.subscribe("Cart",
-  Session.get("sessionId"),
-  Meteor.userId()
-);
-// detect when a cart has been deleted
-// resubscribe will force cart to be rebuilt
-let cart = ReactionCore.Collections.Cart.find();
-cart.observeChanges({
-  removed: function () {
-    Meteor.subscribe("Cart", Session.get("sessionId"), Meteor.userId());
-  }
+// @see http://guide.meteor.com/data-loading.html#changing-arguments
+Tracker.autorun(() => {
+  let sessionId;
+  // we really don't need to track the sessionId here
+  Tracker.nonreactive(() => {
```

Comment:
In this part we do not need to track `SessionId`, because the one thing we need to track here is a `userId`. But we need to send `sessionId` down to server methods. If we not put Session.get("SessionId") in nonreactive, it will be cause of rerun cart subscription on every page refresh.